### PR TITLE
MM-644 Adds the ability to use the arrow keys to scroll the @ mention list and use tab to complete

### PR DESCRIPTION
**Pull request by rgarmsen2295**
This adds the ability to use one's arrow keys to scroll through the @ mention list in the center pane or the RHS along with being able to use tab to complete a mention.

### CODE REVIEW

File: web/react/components/mention_list.jsx [hunk 1]
```diff
@@ -17,14 +17,39 @@ module.exports = React.createClass({
...
-                    self.addFirstMention();
+
+                    var tempSelectedMention = -1;
+                    self.refs['mention' + self.state.selectedMention].deselect();
+                    if (e.which === 38) {    
+                        if (self.getSelection(self.state.selectedMention - 1))
+                            self.setState({ selectedMention: self.state.selectedMention - 1, selectedUsername: self.refs['mention' + (self.state.selectedMention - 1)].props.username });
+                        else {
+                            while (self.getSelection(++tempSelectedMention))
+                                ; //Need to find the top of the list
```
Comment (jwilander) [inline web/react/components/mention_list.jsx]: Syntax looks a little weird here

File: web/react/components/mention_list.jsx [hunk 2]
```diff
@@ -118,17 +213,21 @@ module.exports = React.createClass({
...
-                mentions[i+1] = (
+                mentions[index] = (
                     <Mention
                         ref={'mention' + index}
                         username={users[i].username}
                         secondary_text={users[i].secondary_text}
                         id={users[i].id}
+                        listId={index}
+                        isFocus={this.state.selectedMention === index ? true : false}
+                        handleMouseEnter={this.handleMouseEnter}
```
Comment (jwilander) [inline web/react/components/mention_list.jsx]: If you do this instead:

```handleMouseEnter={function(value) { return function() { self.handleMouseEnter(value); } }(index)}```

Then you shouldn't need to pass in listId to the mention control

**Comment by jwilander** [comment]
Is there a reason why we track selectedMention and selectedUsername in the state? I haven't done too much in depth thinking about it but I feel like we should be able to just track the index of the selected mention in the state (reset it to 0 when the mention list changes).

Also why not just use the props.isFocused all the time instead of adding select()/deselect() functions to mention.jsx if all that's doing is adding a CSS class? Basically just set the mentions_list state with the new selected index whenever it changes and then in the render function when looping through the mentions just update isFocused

File: web/react/components/mention_list.jsx (hunk 1 above)
Comment (rgarmsen2295) [inline web/react/components/mention_list.jsx]: By weird do you mean move the semi-colon and comment up to be on the same line as the while loop?  I don't mind either way, I've just been taught in the past that if you do a "empty" loop that you should do it on a new line for clarity, but 0/5 personally.

File: web/react/components/mention_list.jsx (hunk 2 above)
Comment (rgarmsen2295) [inline web/react/components/mention_list.jsx]: Ok cool, yeah that makes a lot more sense, thanks.

File: web/react/components/mention_list.jsx (hunk 1 above)
Comment (jwilander) [inline web/react/components/mention_list.jsx]: No it's fine the way it is, I'm 0/5 as well I'm just not used to seeing it that way. When we have a standards guide we can decide how we want to do it

**Comment by rgarmsen2295** [comment]
My reasoning behind tracking both was so that if you were on a name, say the 10th list "C" name in a list and then continued to type the name as the list shrunk it would "jump" downwards due to the mention list length shrinking but the selected index remaining the same, so I wanted a way to maintain where you were even as the list shrunk (however backspacing sends you back to mention 0 as that felt better while I was testing) hence the while loop in "componentDidUpdate" so if the mention number and username don't match up it can find the real mention again.

However I think you're right with your second suggestion, how I was doing things changed a lot while I was working on this and thinking about this a day removed it does look pretty redundant.  (still have to test throughly though to be sure, might have had an amazing reason I forgot about) thanks for the great feedback, let me work with it a little more and see what I can do to clean it up!

File: web/react/components/mention_list.jsx (hunk 1 above)
Comment (rgarmsen2295) [inline web/react/components/mention_list.jsx]: Ok, I'll leave it how it is for now and come back to it when we have a guide up and running if necessary.

**Comment by jwilander** [comment]
Sounds good for both

**Comment by rgarmsen2295** [comment]
Ok made and tested the changes, let me know if there's anything else you see worth mentioning

**Pull request merged**